File: wasm/sample/src/xplpc/serializer/json-serializer.ts
import { DecodedRequest } from "../message/decoded-request";
import { Message } from "../message/message";
import { Param } from "../message/param";
import { ExceptionMessage } from "../util/exception-message";
import { Log } from "../util/log";
import { Serializer } from "./serializer";

interface JsonParseContext {
    source?: string;
}

// The runtime hands the reviver the original text, which the standard library types do not describe yet.
declare global {
    interface JSON {
        parse(
            text: string,
            reviver: (
                this: unknown,
                key: string,
                value: unknown,
                context?: JsonParseContext,
            ) => unknown,
        ): unknown;
    }
}

// A javascript number is a double, so an integer literal above the safe range is rounded on parse.
// The reviver sees the original text and rejects the value instead of handing back a different one.
function rejectRoundedIntegers(
    key: string,
    value: unknown,
    context?: JsonParseContext,
): unknown {
    const source = context?.source;

    if (
        typeof value !== "number" ||
        source === undefined ||
        !/^-?\d+$/.test(source) ||
        Number.isSafeInteger(value)
    ) {
        return value;
    }

    throw new RangeError(
        `The value ${source} of "${key}" needs more precision than a javascript number holds`,
    );
}

interface RequestData {
    f?: string;
    p?: { n: string; v: unknown }[];
}

interface ResponseData {
    r?: unknown;
}

function parse<T>(data: string): T {
    return JSON.parse(data, rejectRoundedIntegers) as T;
}

export class JsonSerializer implements Serializer {
    decodeRequest(data: string): DecodedRequest | undefined {
        try {
            const decodedData = parse<RequestData>(data);
            const message = new Message();

            decodedData.p?.forEach((param) => {
                message.set(param.n, param.v);
            });

            return new DecodedRequest(decodedData.f ?? "", message);
        } catch (e: unknown) {
            Log.e("[JsonSerializer : decodeRequest] Error when decode request");
            Log.d(
                "[JsonSerializer : decodeRequest] Error when decode request: " +
                    ExceptionMessage.get(e),
            );
        }

        return undefined;
    }

    encodeFunctionReturnValue(data: unknown): string {
        try {
            return JSON.stringify({
                r: data,
            });
        } catch (e: unknown) {
            Log.e(
                "[JsonSerializer : encodeFunctionReturnValue] Error when encode data",
            );
            Log.d(
                "[JsonSerializer : encodeFunctionReturnValue] Error when encode data: " +
                    ExceptionMessage.get(e),
            );
        }

        return "";
    }

    decodeFunctionReturnValue<T>(data: string): T | undefined {
        // An empty response is the empty value every failing path answers with, not a document that failed to parse.
        if (!data) {
            return undefined;
        }

        try {
            return parse<ResponseData>(data).r as T | undefined;
        } catch (e: unknown) {
            Log.e(
                "[JsonSerializer : decodeFunctionReturnValue] Error when parse json",
            );
            Log.d(
                "[JsonSerializer : decodeFunctionReturnValue] Error when parse json: " +
                    ExceptionMessage.get(e),
            );
        }

        return undefined;
    }

    encodeRequest(functionName: string, ...params: Param[]): string {
        try {
            return JSON.stringify({
                f: functionName,
                p: params.map((item) => {
                    return { n: item.name, v: item.value };
                }),
            });
        } catch (e: unknown) {
            Log.e("[JsonSerializer : encodeRequest] Error when encode data");
            Log.d(
                "[JsonSerializer : encodeRequest] Error when encode data: " +
                    ExceptionMessage.get(e),
            );
        }

        return "";
    }
}

File: python/lib/src/xplpc/serializer/json_serializer.py
import inspect
import json

from xplpc.message.decoded_request import DecodedRequest
from xplpc.message.message import Message
from xplpc.serializer.base_serializer import BaseSerializer
from xplpc.util.log import Log

# The reference format is what the c++ core writes, and python is the only json library here whose defaults differ from it, by spacing the output and by escaping every character outside ascii.
# The wire has no token for a number that is not finite, so writing one would put a document on it that no other bridge can read.
SEPARATORS = (",", ":")


class JsonSerializer(BaseSerializer):
    def decode_function_return_value(self, data, class_type=None):
        # An empty response is the empty value every failing path answers with, not a document that failed to parse.
        if not data:
            return None

        try:
            # A document that carries no value answers the empty one, the same as every other bridge, rather than raising.
            value = json.loads(data).get("r")

            if not class_type:
                return value

            if isinstance(value, dict):
                if inspect.signature(class_type.__init__) == inspect.signature(
                    object.__init__
                ):
                    instance = class_type.__new__(class_type)
                else:
                    instance = class_type()

                instance.__dict__.update(value)
                return instance

            # A value that is not an object cannot fill one, so it answers only when it already is the requested type.
            if isinstance(value, class_type):
                return value

            Log.e(
                "[JsonSerializer : decode_function_return_value] Value has another type"
            )
            Log.d(
                f"[JsonSerializer : decode_function_return_value] Value has another type: {type(value).__name__}"
            )

            return None
        except Exception as e:
            Log.e(
                "[JsonSerializer : decode_function_return_value] Error when parse json"
            )
            Log.d(
                f"[JsonSerializer : decode_function_return_value] Error when parse json: {e}"
            )

        return None

    def decode_request(self, data):
        try:
            decoded_data = json.loads(data)
            message = Message()

            # A parameter that arrived without a value is the same as one carrying null, which is what every other bridge reads it as.
            for param in decoded_data.get("p") or []:
                message.set(param["n"], param.get("v"))

            # The decoded shape has to say what actually arrived, since a name that is not text reaches the proxy and raises there instead.
            function_name = decoded_data.get("f")

            if not isinstance(function_name, str):
                function_name = ""

            return DecodedRequest(function_name, message)
        except Exception as e:
            Log.e("[JsonSerializer : decode_request] Error when decode request")
            Log.d(f"[JsonSerializer : decode_request] Error when decode request: {e}")

        return None

    def encode_function_return_value(self, data):
        try:
            return json.dumps(
                {"r": data},
                separators=SEPARATORS,
                ensure_ascii=False,
                allow_nan=False,
            )
        except Exception as e:
            Log.e(
                "[JsonSerializer : encode_function_return_value] Error when encode data"
            )
            Log.d(
                f"[JsonSerializer : encode_function_return_value] Error when encode data: {e}"
            )

        return ""

    def encode_request(self, function_name, params=None):
        try:
            if not params:
                params = []
            else:
                params = [
                    param.to_json() if hasattr(param, "to_json") else param
                    for param in params
                ]

            return json.dumps(
                {
                    "f": function_name,
                    "p": params,
                },
                separators=SEPARATORS,
                ensure_ascii=False,
                allow_nan=False,
            )
        except Exception as e:
            Log.e("[JsonSerializer : encode_request] Error when encode data")
            Log.d(f"[JsonSerializer : encode_request] Error when encode data: {e}")

        return ""
